### DISCUSSION TITLE
How do I change the keyboard shortcut for autosuggestions from right arrow key to something else?

### DISCUSSION THREAD

**Discussion by Flimm** [xonsh/xonsh#4687] [Q&A] [answered]
When typing a command in xonsh, xonsh may suggest a completion for that entire command if it resembles a command in history. Fish has a similar feature called "autosuggestions". Xonsh has this feature, but I'm not sure what it is called, I couldn't find documentation for it.

![Screenshot](https://user-images.githubusercontent.com/355454/156338910-1022888d-021b-4f56-8a66-87b0b8bdc700.png)

To use the autosuggestion, the user needs to press the right arrow key <kbd>→</kbd>, and then press enter. I would like to change this keyboard shortcut to be something like <kbd>control</kbd>-<kbd>S</kbd> or something like that, because I don't want to reach all the way to the arrow keys. How do I do that?

**Comment by gforsyth** [comment, marked as answer]
`ctrl-e` is also available in addition to right arrow.

If that doesn't work for you I can point you at some resources for overriding key bindings

**Reply by leonplanken** [reply]
Thank you, `ctrl-E` is a good workaround. If I'd want to set another keybinding, how would I go about this?

Also, where in the source is this defined? Searching for `Keys.ControlE` or `Keys.Right` doesn't seem to give any appropriate results.

**Reply by leonplanken** [reply]
Self-replying-- I just found this old discussion: https://groups.google.com/g/xonsh/c/RWyvShZNibg
I'll see if I can get some further traction myself, but any other insights are also appreciated.

**Reply by gforsyth** [reply]
Any keybindings not defined in `xonsh/ptk_shell/key_bindings.py` are upstream defaults -- the upstream "accept suggestion" keybinding is defined here: https://github.com/prompt-toolkit/python-prompt-toolkit/blob/65c3d0607c69c19d80abb052a18569a2546280e5/src/prompt_toolkit/key_binding/bindings/auto_suggest.py#L38-L49

(note that `"c-e"` and `keys.ControlE` are synonymous).

If you want to add an additional key to handle that action, you can follow the tutorial here for how to define them inside of your `xonshrc` -- https://xon.sh/tutorial_ptk.html

**Comment by Spitfire1900** [comment]
Related to this, it looks like `Ctrl+->` is only available in Xonsh when using emacs key bindings, anyone know if its available in VI_MODE?

Related discussion: https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1859

**Reply by gforsyth** [reply]
Yeah, control + right arrow doesn't work in VI_MODE, but, hilariously, Ctrl + e does.

**Reply by Spitfire1900** [reply]
I _almost_ see have something here in my .xonshrc binding accept auto-completion with `Ctrl-space`
I do have a bug though,  see `# BUG:` most of the way down.


```python
### Custom key bindings
from prompt_toolkit.keys import Keys
from prompt_toolkit.key_binding.key_processor import KeyPressEvent
from prompt_toolkit.filters import Condition
from prompt_toolkit.key_binding.key_bindings import KeyBindings

@events.on_ptk_create
def custom_keybindings(bindings, **kw):

    ## Accept partial completion
    ##     derived from https://github.com/prompt-toolkit/python-prompt-toolkit/blob/465ab02854763fafc0099a2e38a56328c1cb0625/src/prompt_toolkit/key_binding/bindings/auto_suggest.py#L54
    @Condition
    def suggestion_available() -> bool:
    
        from prompt_toolkit.application.current import get_app
        app = get_app()
        return (
            app.current_buffer.suggestion is not None
            and len(app.current_buffer.suggestion.text) > 0
            and app.current_buffer.document.is_cursor_at_the_end
        )
    
    E = KeyPressEvent
    
    @bindings.add('c-space', filter=suggestion_available) # BUG: suggestion_available -> False causes hang until Ctrl+C
    def _fill(event: E) -> None:
        """
        Fill partial suggestion.
        """
        b = event.current_buffer
        suggestion = b.suggestion

        if suggestion:
            t = re.split(r"([^\s/]+(?:\s+|/))", suggestion.text)
            b.insert_text(next(x for x in t if x))
```

**Reply by gforsyth** [reply]
there's a built-in `has_suggestion` condition in `prompt_toolkit.filters` that might help

**Reply by gforsyth** [reply]
and I think the action you want is also predefined as `forward_word`: https://github.com/prompt-toolkit/python-prompt-toolkit/blob/master/src/prompt_toolkit/key_binding/bindings/named_commands.py#L125-L134

**Reply by Spitfire1900** [reply]
Binding this to 'c-right' instead of 'c-space' avoided hangs.

**Comment by Spitfire1900** [comment]
> and I think the action you want is also predefined as `forward_word`: https://github.com/prompt-toolkit/python-prompt-toolkit/blob/master/src/prompt_toolkit/key_binding/bindings/named_commands.py#L125-L134

This does not work the same as [_fill()](https://github.com/prompt-toolkit/python-prompt-toolkit/blob/465ab02854763fafc0099a2e38a56328c1cb0625/src/prompt_toolkit/key_binding/bindings/auto_suggest.py#L54) and prompt_toolkit's `_fill()` does not seem exposed in a way to bind to it directly.

**Reply by Spitfire1900** [reply]
Final implementation to get `c-right` accepting autosuggestion in VI_MODE. This also lets you bind any other key-binding.


```python
### Custom key bindings
from prompt_toolkit.keys import Keys
from prompt_toolkit.key_binding.key_processor import KeyPressEvent
from prompt_toolkit.filters import has_suggestion

@events.on_ptk_create
def custom_keybindings(bindings, **kw):
    
    @bindings.add(Keys.ControlRight, filter=has_suggestion)
    def _fill(event: KeyPressEvent) -> None:
        """
        Fill partial suggestion.
            derived from:
            https://github.com/prompt-toolkit/python-prompt-toolkit/blob/465ab02854763fafc0099a2e38a56328c1cb0625/src/prompt_toolkit/key_binding/bindings/auto_suggest.py#L54
        """
        
        
        b = event.current_buffer
        suggestion = b.suggestion

        if suggestion:
            t = re.split(r"([^\s/]+(?:\s+|/))", suggestion.text)
            b.insert_text(next(x for x in t if x))
```
